### DISCUSSION TITLE
Quasar table - Expanding rows

### DISCUSSION THREAD

**Discussion by pythonormen** [zauberzeug/nicegui#574] [Q&A] [answered]
Thank you for the new release 👍 

Would anyone mind to provide a simple example on how to use Expanding rows with Nicegui and the new table element?

<img width="867" alt="image" src="https://user-images.githubusercontent.com/125542430/226411287-51db102a-fce2-4e76-8ba8-9f835755fc97.png">

https://quasar.dev/vue-components/table#example--internal-expansion-model

**Comment by falkoschindler** [comment, marked as answer]
Although it's definitely not simple, at least it's possible thanks to NiceGUI's new support for scoped slots:

```py
columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'age', 'label': 'Age', 'field': 'age'},
]
rows = [
    {'name': 'Alice', 'age': 18},
    {'name': 'Bob', 'age': 21},
    {'name': 'Carol'},
]

with ui.table(columns=columns, rows=rows, row_key='name').classes('w-96') as table:
    table.add_slot('header', r'''
        <q-tr :props="props">
            <q-th auto-width />
            <q-th v-for="col in props.cols" :key="col.name" :props="props">
                {{ col.label }}
            </q-th>
        </q-tr>
    ''')
    table.add_slot('body', r'''
        <q-tr :props="props">
            <q-td auto-width>
                <q-btn size="sm" color="accent" round dense @click="props.expand = !props.expand" :icon="props.expand ? 'remove' : 'add'" />
            </q-td>
            <q-td v-for="col in props.cols" :key="col.name" :props="props">
                {{ col.value }}
            </q-td>
        </q-tr>
        <q-tr v-show="props.expand" :props="props">
            <q-td colspan="100%">
                <div class="text-left">This is expand slot for row above: {{ props.row.name }}.</div>
            </q-td>
        </q-tr>
    ''')
```

I basically copy-pasted the templates from the example at https://quasar.dev/vue-components/table#expanding-rows.

**Reply by pythonormen** [reply]
Wow, thank you. Way to complex for me, but will add it as another milestone.

1) Just a short one on this. Would it possible to add information from some random source/variable? 
The idea is that I have my table but can expand and show more details to the end user.

2) Regarding quasar table but not related to expanding rows. When I add selection=single, shouldn't it only be possible to select single item?
I.e. so it's similar to this one https://quasar.dev/vue-components/table#example--single-selection or is it intended that you can select multiple instead of clearing?

    ```py
    from nicegui import ui

    columns = [
        {'name': 'name', 'label': 'Name', 'field': 'name', 'required': True, 'align': 'left'},
        {'name': 'age', 'label': 'Age', 'field': 'age', 'sortable': True},
    ]
    rows = [
        {'name': 'Alice', 'age': 18},
        {'name': 'Bob', 'age': 21},
        {'name': 'Carol'},
    ]
    ui.table(columns=columns, rows=rows, row_key='name', selection='single')

    ui.run()
    ```

3) And on the same topic as 2. Is there any way to clear the selection?
Maybe I'm doing it the wrong way by only updating the table from the source (sqlite db) and then using table.update() to refresh the content (perhaps the table and db should be modified separately). But anyway. In my setup in case a selected row has been removed in the database and a user has that row selected, then the selection is kept even though the row is gone.

Don't know if this is a bug or not. Well my question is non the less if there is a way to clear the selection?

**Reply by falkoschindler** [reply]
Let me answer your questions one by one:

1. You can extend the `rows` list with more fields like

    ```py
    rows = [
        {'name': 'Alice', 'age': 18, 'address': '123 Main St.'},
        {'name': 'Bob', 'age': 21, 'address': '456 Main St.'},
        {'name': 'Carol', 'age': 19, 'address': '789 Main St.'},
    ]
    ```

    and refer to them within the expansion like

    ```html
    <div class="text-left">{{ props.row.name }} lives at {{ props.row.address }}.</div>
    ```

2. That's definitely a bug. Thanks for bringing this up! I opened a new issue: #609.

3. You can clear the selection as follows:

    ```py
    def clear_selection():
        table.selected[:] = []
        table.update()

    ui.button('Clear selection', on_click=clear_selection)
    ```
    
    Make sure not to change the reference of `table.selected`. Therefore I'm using `[:]` here.
